🐛Prepends heading to amp-story-bookend components if first is not a heading

Enriqe opened the pull request:
Fixes #15089

gmajoulet commented on extensions/amp-story/1.0/bookend/bookend-component.js:
```diff
@@ -102,6 +105,29 @@ function componentBuilderInstanceFor(type) {
+              .buildElement({'text': title}, doc), container.firstChild);
+    });
+  }
+}
```
From what I understand, `buildElements` will return the `DocumentFragment` BEFORE the heading section is prepended. It's very confusing to receive some DOM from `buildElements` that is going to be updated later, at the end of the frame. It makes things hard to predict, and therefore makes it very easy to introduce bugs when using this function.
The code running in the callback of the `Services.localizationServiceForOrNull(win)` promise will run in a microtask, after all the synchronous code is executed. Timeline would look like:

1. some call to `buildElements`
2. call to prependTitle
3. Components get built
4. `buildElements` return the components
5. Someone could do something with these returned components, assuming the heading is present
6. Heading actually gets prepended (should have been in (2))

We should find a way for `buildElements` to return the final DOM elements.

Also, I don't think we should duplicate the logic to build a component (here, and in `buildElements`).

Could we simply check the components list, prepend a `heading` component, if needed, in the `components` array, and keep the same building logic?

Changing `buildElements` to return a `Promise` (if you need to, because of this localization service) should be straightforward to implement, and make this function easier to predict / use.

gmajoulet commented:
From what I understand, `buildElements` will return the `DocumentFragment` BEFORE the heading section is prepended. It's very confusing to receive some DOM from `buildElements` that is going to be updated later, at the end of the frame. It makes things hard to predict, and therefore makes it very easy to introduce bugs when using this function.
The code running in the callback of the `Services.localizationServiceForOrNull(win)` promise will run in a microtask, after all the synchronous code is executed. Timeline would look like:

1. some call to `buildElements`
2. call to prependTitle
3. Components get built
4. `buildElements` return the components
5. Someone could do something with these returned components, assuming the heading is present
6. Heading actually gets prepended (should have been in (2))

We should find a way for `buildElements` to return the final DOM elements.

Also, I don't think we should duplicate the logic to build a component (here, and in `buildElements`).

Could we simply check the components list, prepend a `heading` component, if needed, in the `components` array, and keep the same building logic?

Changing `buildElements` to return a `Promise` (if you need to, because of this localization service) should be straightforward to implement, and make this function easier to predict / use.

gmajoulet commented on extensions/amp-story/1.0/bookend/amp-story-bookend.js:
```diff
@@ -495,6 +495,7 @@ export class AmpStoryBookend extends AMP.BaseElement {
 
   /**
    * @param {!./bookend-component.BookendDataDef} bookendConfig
+   * @return {!Promise<?../localization.LocalizationService>|undefined}
```
Should this method return the rendered bookend instead? Not sure it's super useful to get the `LocalizationService` as a response of the `renderBookend_` method.

gmajoulet commented on extensions/amp-story/1.0/bookend/amp-story-bookend.js:
```diff
@@ -505,21 +506,31 @@ export class AmpStoryBookend extends AMP.BaseElement {
+              BookendComponent.buildContainer(this.getInnerContainer_(),
+                  this.win.document));
+          this.mutateElement(() => container.appendChild(bookendEls));
+          return localizationService;
```
Re: above comment
maybe we could simply return `bookendEls` instead here (or return nothing, if we don't have any use for it)

Enriqe commented:
@gmajoulet @newmuis 

We need `renderBookend_` to return a promise to ensure that `loadConfigAndMaybeRenderBookend` renders the bookend first before returning the config. This is something I needed to change in order to make unit tests work. But I agree that it's not very relevant for `renderBookend` to return a promise that resolves a `localizationService`, so I changed it to only return a promise (since we don't actually used the resolved value of the promise anywhere), and updated the JSDoc as well.
